TST: `optimize.elementwise.bracket_minimum`: CuPy failure

cc @j-bowhay @mdhaber (on the bright side, this is the only current failure for `pixi run test-cuda` :tada:)

```
_________________________________________________________________________ TestBracketMinimum.test_flags[cupy] __________________________________________________________________________
scipy/optimize/tests/test_bracket.py:485: in test_flags
    result = _bracket_minimum(f, xm0, xl0=xl0, xr0=xr0, xmin=xmin,
        args       = (array([0, 1, 2, 3, 4]),)
        f          = <function TestBracketMinimum.test_flags.<locals>.f at 0x7f636318d800>
        self       = <scipy.optimize.tests.test_bracket.TestBracketMinimum object at 0x7f641e915b20>
        xl0        = array([-1., -1., -1., -1.,  6.])
        xm0        = array([0., 0., 0., 0., 4.])
        xmin       = array([-inf,  -1., -inf, -inf,   8.])
        xp         = <module 'cupy' from '/home/lucas/dev/pixi-dev-scipystack/scipy/.pixi/envs/array-api-cuda/lib/python3.12/site-packages/cupy/__init__.py'>
        xr0        = array([1., 1., 1., 1., 2.])
scipy/optimize/tests/test_bracket.py:36: in _bracket_minimum
    res = bracket_minimum(*args, **kwargs)
        args       = (<function TestBracketMinimum.test_flags.<locals>.f at 0x7f636318d800>, array([0., 0., 0., 0., 4.]))
        kwargs     = {'args': (array([0, 1, 2, 3, 4]),), 'maxiter': 3, 'xl0': array([-1., -1., -1., -1.,  6.]), 'xmin': array([-inf,  -1., -inf, -inf,   8.]), ...}
scipy/optimize/_elementwise.py:791: in bracket_minimum
    res = _bracket_minimum(f, xm0, xl0=xl0, xr0=xr0, xmin=xmin, xmax=xmax,
        args       = (array([0, 1, 2, 3, 4]),)
        f          = <function TestBracketMinimum.test_flags.<locals>.f at 0x7f636318d800>
        factor     = None
        maxiter    = 3
        xl0        = array([-1., -1., -1., -1.,  6.])
        xm0        = array([0., 0., 0., 0., 4.])
        xmax       = None
        xmin       = array([-inf,  -1., -inf, -inf,   8.])
        xr0        = array([1., 1., 1., 1., 2.])
scipy/optimize/_bracket.py:583: in _bracket_minimum
    temp = eim._initialize(func, xs, args)
        args       = (array([0, 1, 2, 3, 4]),)
        callback   = None
        factor     = array([2., 2., 2., 2., 2.])
        func       = <function TestBracketMinimum.test_flags.<locals>.f at 0x7f636318d800>
        maxiter    = array(3)
        temp       = (<function TestBracketMinimum.test_flags.<locals>.f at 0x7f636318d800>, array([0., 0., 0., 0., 4.]), array([-1., -1., ... -1.,  6.]), array([1., 1., 1., 1., 2.]), array([-inf,  -1., -inf, -inf,   8.]), array([inf, inf, inf, inf, inf]), ...)
        xl0        = array([-1., -1., -1., -1.,  6.])
        xm0        = array([0., 0., 0., 0., 4.])
        xmax       = array([inf, inf, inf, inf, inf])
        xmin       = array([-inf,  -1., -inf, -inf,   8.])
        xp         = <module 'scipy._lib.array_api_compat.cupy' from '/home/lucas/dev/pixi-dev-scipystack/scipy/scipy/build-install/lib/python3.12/site-packages/scipy/_lib/array_api_compat/cupy/__init__.py'>
        xr0        = array([1., 1., 1., 1., 2.])
        xs         = (array([-1., -1., -1., -1.,  6.]), array([0., 0., 0., 0., 4.]), array([1., 1., 1., 1., 2.]))
scipy/_lib/_elementwise_iterative_method.py:90: in _initialize
    fs = [xp.asarray(func(x, *args)) for x in xs]
        args       = [array([0, 1, 2, 3, 4])]
        complex_ok = False
        func       = <function TestBracketMinimum.test_flags.<locals>.f at 0x7f636318d800>
        nx         = 3
        preserve_shape = None
        xas        = [array([-1., -1., -1., -1.,  6.]), array([0., 0., 0., 0., 4.]), array([1., 1., 1., 1., 2.]), array([0, 1, 2, 3, 4])]
        xat        = dtype('float64')
        xp         = <module 'scipy._lib.array_api_compat.cupy' from '/home/lucas/dev/pixi-dev-scipystack/scipy/scipy/build-install/lib/python3.12/site-packages/scipy/_lib/array_api_compat/cupy/__init__.py'>
        xs         = [array([-1., -1., -1., -1.,  6.]), array([0., 0., 0., 0., 4.]), array([1., 1., 1., 1., 2.])]
scipy/optimize/tests/test_bracket.py:477: in f
    return [funcs[j](x) for x, j in zip(xs, js)]
E   TypeError: list indices must be integers or slices, not ndarray
        funcs      = [<function TestBracketMinimum.test_flags.<locals>.f.<locals>.<lambda> at 0x7f636318fba0>, <function TestBracketMinimum...>.<lambda> at 0x7f636318d260>, <function TestBracketMinimum.test_flags.<locals>.f.<locals>.<lambda> at 0x7f636318cf40>]
        js         = array([0, 1, 2, 3, 4])
        xp         = <module 'cupy' from '/home/lucas/dev/pixi-dev-scipystack/scipy/.pixi/envs/array-api-cuda/lib/python3.12/site-packages/cupy/__init__.py'>
        xs         = array([-1., -1., -1., -1.,  6.])
```

that fixes that error but gives a new one:

```
_________________________________________________________________________ TestBracketMinimum.test_flags[cupy] __________________________________________________________________________
scipy/optimize/tests/test_bracket.py:485: in test_flags
    result = _bracket_minimum(f, xm0, xl0=xl0, xr0=xr0, xmin=xmin,
        args       = (array([0, 1, 2, 3, 4]),)
        f          = <function TestBracketMinimum.test_flags.<locals>.f at 0x7e98d3ff5800>
        self       = <scipy.optimize.tests.test_bracket.TestBracketMinimum object at 0x7e98a304c3e0>
        xl0        = array([-1., -1., -1., -1.,  6.])
        xm0        = array([0., 0., 0., 0., 4.])
        xmin       = array([-inf,  -1., -inf, -inf,   8.])
        xp         = <module 'cupy' from '/home/lucas/dev/pixi-dev-scipystack/scipy/.pixi/envs/cupy/lib/python3.12/site-packages/cupy/__init__.py'>
        xr0        = array([1., 1., 1., 1., 2.])
scipy/optimize/tests/test_bracket.py:36: in _bracket_minimum
    res = bracket_minimum(*args, **kwargs)
        args       = (<function TestBracketMinimum.test_flags.<locals>.f at 0x7e98d3ff5800>, array([0., 0., 0., 0., 4.]))
        kwargs     = {'args': (array([0, 1, 2, 3, 4]),), 'maxiter': 3, 'xl0': array([-1., -1., -1., -1.,  6.]), 'xmin': array([-inf,  -1., -inf, -inf,   8.]), ...}
scipy/optimize/_elementwise.py:791: in bracket_minimum
    res = _bracket_minimum(f, xm0, xl0=xl0, xr0=xr0, xmin=xmin, xmax=xmax,
        args       = (array([0, 1, 2, 3, 4]),)
        f          = <function TestBracketMinimum.test_flags.<locals>.f at 0x7e98d3ff5800>
        factor     = None
        maxiter    = 3
        xl0        = array([-1., -1., -1., -1.,  6.])
        xm0        = array([0., 0., 0., 0., 4.])
        xmax       = None
        xmin       = array([-inf,  -1., -inf, -inf,   8.])
        xr0        = array([1., 1., 1., 1., 2.])
scipy/optimize/_bracket.py:583: in _bracket_minimum
    temp = eim._initialize(func, xs, args)
        args       = (array([0, 1, 2, 3, 4]),)
        callback   = None
        factor     = array([2., 2., 2., 2., 2.])
        func       = <function TestBracketMinimum.test_flags.<locals>.f at 0x7e98d3ff5800>
        maxiter    = array(3)
        temp       = (<function TestBracketMinimum.test_flags.<locals>.f at 0x7e98d3ff5800>, array([0., 0., 0., 0., 4.]), array([-1., -1., ... -1.,  6.]), array([1., 1., 1., 1., 2.]), array([-inf,  -1., -inf, -inf,   8.]), array([inf, inf, inf, inf, inf]), ...)
        xl0        = array([-1., -1., -1., -1.,  6.])
        xm0        = array([0., 0., 0., 0., 4.])
        xmax       = array([inf, inf, inf, inf, inf])
        xmin       = array([-inf,  -1., -inf, -inf,   8.])
        xp         = <module 'scipy._lib.array_api_compat.cupy' from '/home/lucas/dev/pixi-dev-scipystack/scipy/scipy/build-install/lib/python3.12/site-packages/scipy/_lib/array_api_compat/cupy/__init__.py'>
        xr0        = array([1., 1., 1., 1., 2.])
        xs         = (array([-1., -1., -1., -1.,  6.]), array([0., 0., 0., 0., 4.]), array([1., 1., 1., 1., 2.]))
scipy/_lib/_elementwise_iterative_method.py:90: in _initialize
    fs = [xp.asarray(func(x, *args)) for x in xs]
        args       = [array([0, 1, 2, 3, 4])]
        complex_ok = False
        func       = <function TestBracketMinimum.test_flags.<locals>.f at 0x7e98d3ff5800>
        nx         = 3
        preserve_shape = None
        xas        = [array([-1., -1., -1., -1.,  6.]), array([0., 0., 0., 0., 4.]), array([1., 1., 1., 1., 2.]), array([0, 1, 2, 3, 4])]
        xat        = dtype('float64')
        xp         = <module 'scipy._lib.array_api_compat.cupy' from '/home/lucas/dev/pixi-dev-scipystack/scipy/scipy/build-install/lib/python3.12/site-packages/scipy/_lib/array_api_compat/cupy/__init__.py'>
        xs         = [array([-1., -1., -1., -1.,  6.]), array([0., 0., 0., 0., 4.]), array([1., 1., 1., 1., 2.])]
scipy/_lib/array_api_compat/cupy/_aliases.py:111: in asarray
    return cp.array(obj, dtype=dtype, **kwargs)
        copy       = <object object at 0x7e98d5a8d880>
        device     = None
        dtype      = None
        kwargs     = {}
        obj        = [array(6.25), array(-1.), array(-1.), nan, array(36.)]
../../../../../.pixi/envs/cupy/lib/python3.12/site-packages/cupy/_creation/from_data.py:53: in array
    return _core.array(obj, dtype, copy, order, subok, ndmin, blocking)
        blocking   = False
        copy       = True
        dtype      = None
        ndmin      = 0
        obj        = [array(6.25), array(-1.), array(-1.), nan, array(36.)]
        order      = 'K'
        subok      = False
cupy/_core/core.pyx:2408: in cupy._core.core.array
    ???
cupy/_core/core.pyx:2435: in cupy._core.core.array
    ???
cupy/_core/core.pyx:2571: in cupy._core.core._array_default
    ???
cupy/_core/core.pyx:1481: in cupy._core.core._ndarray_base.__array__
    ???
E   TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
```

--- a/scipy/optimize/tests/test_bracket.py
+++ b/scipy/optimize/tests/test_bracket.py
@@ -471,10 +471,10 @@ class TestBracketMinimum:
             funcs = [lambda x: (x - 1.5)**2,
                      lambda x: x,
                      lambda x: x,
-                     lambda x: xp.nan,
+                     lambda x: xp.asarray(xp.nan),
                      lambda x: x**2]
 
-            return [funcs[j](x) for x, j in zip(xs, js)]
+            return [funcs[int(j)](x) for x, j in zip(xs, js)]
 
         args = (xp.arange(5, dtype=xp.int64),)
         xl0 = xp.asarray([-1.0, -1.0, -1.0, -1.0, 6.0])

